langchain4j-weaviate/src/main/java/dev/langchain4j/store/embedding/WeaviateEmbeddingStoreImpl.java

```diff
@@ -0,0 +1,192 @@
+            Field.builder().name("vector").build()
+          )
+          .build()
+      )
+      .withNearVector(
+        NearVectorArgument
+          .builder()
+          .vector(referenceEmbedding.vectorAsList().toArray(new Float[0]))
+          .certainty((float) minCertainty)
+          .build()
+      )
+      .withLimit(maxResults)
+      .run();
+
+    if (result.hasErrors()) {
+      return emptyList();
+    }
+
+    val resGetPart = ((Map<String, Map>) result.getResult().getData()).entrySet().stream().findFirst();
+    if (!resGetPart.isPresent()) {
+      return emptyList();
+    }
+
+    val resItemsPart = resGetPart.get().getValue().entrySet().stream().findFirst();
+    if (!resItemsPart.isPresent()) {
+      return emptyList();
+    }
+
+    val resItems = ((Map.Entry<String, List<Map<String, ?>>>) resItemsPart.get()).getValue();
+
+    return resItems.stream().map(WeaviateEmbeddingStoreImpl::toEmbeddingMatch).collect(Collectors.toList());
+  }
+
+  private List<String> addAll(List<String> ids, List<Embedding> embeddings, List<TextSegment> embedded) {
+    if (embedded != null && embeddings.size() != embedded.size()) {
+      throw new IllegalArgumentException("The list of embeddings and embedded must have the same size");
+    }
+
+    val resIds = new ArrayList<String>();
+    val objects = new ArrayList<WeaviateObject>();
+    for (int i = 0; i < embeddings.size(); i++) {
+      val id = ids != null ? ids.get(i) : embedded != null ? generateUUI(embedded.get(i).text()) : generateRandomId();
+      resIds.add(id);
+      objects.add(buildObject(id, embeddings.get(i), embedded != null ? embedded.get(i).text() : null));
+    }
+
+    client
+      .batch()
+      .objectsBatcher()
+      .withObjects(objects.toArray(new WeaviateObject[0]))
+      .withConsistencyLevel(ConsistencyLevel.ALL)
+      .run();
+
+    return resIds;
+  }
+
+  private WeaviateObject buildObject(String id, Embedding embedding, String text) {
+    val builder = WeaviateObject
+      .builder()
+      .className(objectClass)
+      .id(id)
+      .vector(embedding.vectorAsList().toArray(new Float[0]));
+
+    if (text != null) {
+      val props = new HashMap<String, Object>();
+      props.put(METADATA_TEXT_SEGMENT, text);
+
+      builder.properties(props);
+    }
+
+    return builder.build();
+  }
+
+  private static EmbeddingMatch<TextSegment> toEmbeddingMatch(Map<String, ?> item) {
+    val additional = (Map<String, ?>) item.get(ADDITIONALS);
+
+    return new EmbeddingMatch<>(
+      (String) additional.get("id"),
+      Embedding.from(
+        ((List<Double>) additional.get("vector")).stream().map(Double::floatValue).collect(Collectors.toList())
+      ),
+      TextSegment.from((String) item.get(METADATA_TEXT_SEGMENT)),
+      (Double) additional.get("certainty")
+    );
+  }
+
+  // TODO this shall be migrated to some common place
+  @SneakyThrows
+  private static String generateUUI(String input) {
```

Comment:
As far as I understand, this method produces UUID from the input string's content? In case user wants to store the same content twice (maybe with different metadata, or same document, but for different end-users), it will not be able to do so? Or old object will nbe rewritten? I would rather generate a content-independent UUID for each embedding, if user did not provide it's own.

Reply:
yes, my idea was to avoid duplications, yes, the old object will be rewritten. Let me experiment with new random ID then....

Reply:
ok, I've player around - of course, randomized ID means new entry in DB. To have a control about it, I've just added the config parameter "avoidDups", added javadoc and updated the example. Pls review.